Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,175 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+struct MissingBytes: Error {
+    
+}
+
+extension ByteBuffer {
+    
+    mutating func parseUnwindingIfNeeded<T>(_ closure: (inout ByteBuffer) throws -> T?) rethrows -> T? {
+        let save = self
+        do {
+            return try closure(&self)
+        } catch is MissingBytes {
```

Comment:
Why are we using `MissingBytes` for this, instead of just having the closure return `nil`?

Reply:
In fact the inner closure _can_ return `nil`. Under what circumstances would those inner closures return `nil` but not mean "we couldn't read enough bytes"?

Reply:
It means the rest of the code can be a lot nicer. I don't need anyway near as many guards, and can just handle the error in one place like this.

Reply:
It seems to me that it only adds one guard: right here, instead of the `catch is MissingBytes`. What am I missing?

Reply:
In the example below I am able to just call `try buffer.readAddressType()` instead of having to guard the result.

```swift
mutating func readServerResponse() throws -> ServerResponse {
        return try self.parseUnwindingIfNeeded { buffer in
            try buffer.readAndValidateProtocolVersion()
            guard let reply = buffer.readInteger(as: UInt8.self).map({ Reply(value: $0) }) else {
                throw SOCKSError.MissingBytes()
            }
            try buffer.readAndValidateReserved()
            let boundAddress = try buffer.readAddressType()
            return .init(reply: reply, boundAddress: boundAddress)
        }
    }
```

Reply:
@Lukasa you may need to refresh, the code has changed a fair bit since your review